std/algebra/emulated/sw_bls12381/hints.go

```diff
@@ -14,7 +14,135 @@ func init() {
 
 // GetHints returns all hint functions used in the package.
 func GetHints() []solver.Hint {
-	return []solver.Hint{finalExpHint}
+	return []solver.Hint{
+		finalExpHint,
+		pairingCheckHint,
+	}
+}
+
+func pairingCheckHint(nativeMod *big.Int, nativeInputs, nativeOutputs []*big.Int) error {
```

Comment:
This method is essentially very similar to the finalExpHint? It seems the only difference is that in this method we compute the Milller loop result ourselves. I would refactor the common part into a separate function.